### PR TITLE
feat: Add NexusViewer

### LINKED ISSUE #1166: NexusViewer
## Overview

Create a new contract viewer to call different contracts of the Nexus platform according to the requirements here: https://docs.google.com/document/d/1MLjGrUcfnUwRsONhXJD-IENw-TEMmdKVM9ayut1ApsQ/edit

### PR DESCRIPTION
**Pull request by rackstar**
## Context

Closes #1166 

Add NexusViewer to provide a unified view of system-wide data from various contracts within the Nexus Mutual platform

NOTE: this PR is behind https://github.com/NexusMutual/smart-contracts/pull/1164

## Changes proposed in this pull request

* Add NexusViewer
* Add mocks
* Add fork tests
* Add unit tests

## Test plan

* Add fork tests
* Add unit tests

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [x] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/modules/viewer/NexusViewer.sol [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+  struct ClaimableNxm {
+    uint governanceRewards;
+    IStakingViewer.AggregatedTokens aggregateStakingTokens;
+    IAssessmentViewer.AssessmentRewards assessmentRewards;
+    LegacyPoolStake legacyPooledStake;
+    uint v1CoverNotesAmount;
+  }
+
+  struct LockedNxm {
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:30]: let's stick to "StakedNXM" as in the UI, as requested by @mixplore

File: contracts/modules/viewer/NexusViewer.sol [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      governanceRewards: governanceRewards,
+      v1CoverNotesAmount: withdrawableAmount
+    });
+  }
+
+  /// @notice Retrieves the locked NXM tokens for a given member.
+  /// @param member The address of the member to query.
+  /// @param tokenIds An array of staking NFT token IDs associated with the member.
+  /// @return A LockedNxm struct containing details of the locked NXM tokens.
+  function getLockedNxm(address member, uint[] calldata tokenIds) public view returns (LockedNxm memory) {
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:76]: rename to `getStakedNxm`

File: contracts/modules/viewer/NexusViewer.sol [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import {IAssessmentViewer} from "../../interfaces/IAssessmentViewer.sol";
+import {IGovernance} from "../../interfaces/IGovernance.sol";
+import {INXMMaster} from "../../interfaces/INXMMaster.sol";
+import {IPooledStaking} from "../../interfaces/IPooledStaking.sol";
+import {IStakingViewer} from "../../interfaces/IStakingViewer.sol";
+import {ITokenController} from "../../interfaces/ITokenController.sol";
+
+/// @title NexusViewer Contract
+/// @notice This contract provides a unified view of system-wide data from various contracts within the Nexus Mutual platform.
+contract NexusViewer is Multicall {
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:16]: for consistency, shouldn't we have an interface for this contract as well?

File: contracts/modules/viewer/NexusViewer.sol [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  struct LegacyPoolStake {
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:17]: struct LegacyPoolStake -> struct LegacyPooledStaking, for correctness

File: contracts/modules/viewer/NexusViewer.sol [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    uint deposit;
+    uint reward;
+  }
+
+  struct ClaimableNxm {
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:22]: this struct is missing assessmentStake, in case there is stake that's not locked

Comment (roxdanila) [reply contracts/modules/viewer/NexusViewer.sol:22]: would be good to leave a comment on each row with what it represents - i.e. 
```
    uint governanceRewards;  // Governance rewards
    IStakingViewer.AggregatedTokens aggregateStakingTokens; // Expired stake deposits + staking rewards
    IAssessmentViewer.AssessmentRewards assessmentRewards; // Assessment rewards
    LegacyPoolStake legacyPooledStake; // Legacy pooled staking deposits + rewards
    uint v1CoverNotesAmount; // Legacy cover note deposits
```

File: contracts/modules/viewer/NexusViewer.sol [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  IAssessmentViewer public immutable assessmentViewer;
+
+  constructor(INXMMaster _master, IStakingViewer _stakingViewer, IAssessmentViewer _assessmentViewer) {
+    master = _master;
+    stakingViewer = _stakingViewer;
+    assessmentViewer = _assessmentViewer;
+  }
+
+  /// @notice Retrieves the claimable NXM tokens for a given member.
+  /// @dev This does not include NXM from manager rewards (use stakingViewer.getManagerTokenRewards)
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:46]: we should include this, why do a separate call for it and have the UI manage this customly?

File: contracts/modules/viewer/NexusViewer.sol [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  /// @param tokenIds An array of staking NFT token IDs associated with the member.
+  /// @return A ClaimableNxm struct containing details of the claimable NXM tokens.
+  function getClaimableNxm(address member, uint[] calldata tokenIds) public view returns (ClaimableNxm memory) {
+
+    IPooledStaking legacyPooledStaking = _legacyPooledStaking();
+    uint deposit = legacyPooledStaking.stakerDeposit(member);
+    uint reward = legacyPooledStaking.stakerReward(member);
+    LegacyPoolStake memory legacyPooledStake = LegacyPoolStake({deposit: deposit, reward: reward});
+
+    IStakingViewer.AggregatedTokens memory aggregatedTokens = stakingViewer.getAggregatedTokens(tokenIds);
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:57]: rename `aggregatedTokens` to `stakingTokens` / `aggregatedStakingTokens` for readability

File: contracts/modules/viewer/NexusViewer.sol [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    IAssessmentViewer.AssessmentRewards memory assessmentRewards = assessmentViewer.getRewards(member);
+
+    uint governanceRewards = _governance().getPendingReward(member);
+    (, , uint withdrawableAmount) = _tokenController().getWithdrawableCoverNotes(member);
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:61]: rename `withdrawableAmount` to `legacyCoverNotes`

File: contracts/modules/viewer/NexusViewer.sol [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    uint deposit;
+    uint reward;
+  }
+
+  struct ClaimableNxm {
+    uint governanceRewards;
+    IStakingViewer.AggregatedTokens aggregateStakingTokens;
+    IAssessmentViewer.AssessmentRewards assessmentRewards;
+    LegacyPoolStake legacyPooledStake;
+    uint v1CoverNotesAmount;
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:27]: rename `v1CoverNotesAmount` to `legacyCoverNotesAmount`

File: contracts/modules/viewer/NexusViewer.sol (hunk 5 above)
Comment (roxdanila) [reply contracts/modules/viewer/NexusViewer.sol:22]: consider splitting `IStakingViewer.AggregatedTokens aggregateStakingTokens; // Expired stake deposits + staking rewards` into 2: `expiredStake` and `rewards`, as only these should be considered when calculating the claimable nxm. this way, you can abstract away some of this cognitive overhead from the FE developers

File: contracts/modules/viewer/NexusViewer.sol [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  struct LockedNxm {
+    IStakingViewer.AggregatedTokens aggregatedTokens;
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:31]: consider only using `activeStake` here from the AggregatedTokens struct. the other 2 fields are irelevant when referring to stakedNXM

File: contracts/modules/viewer/NexusViewer.sol (hunk 5 above)
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:22]: Also missing:
- manager staking rewards
- legacy claim assessment tokens. they can be read via TokenController.tokensLocked(user, formatBytes32String('CLA'))

File: contracts/modules/viewer/NexusViewer.sol (hunk 1 above)
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:30]: watch out that for Assessment rewards, the struct contains 2 values:
`totalPendingAmountInNXM` and `withdrawableAmountInNXM`. I presume you need to use "totalPendingAmountInNXM - withdrawableAmountInNXM` here, and `withdrawableAmountInNXM` above in ClaimableNXM

File: contracts/modules/viewer/NexusViewer.sol [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    stakingViewer = _stakingViewer;
+    assessmentViewer = _assessmentViewer;
+  }
+
+  /// @notice Retrieves the claimable NXM tokens for a given member.
+  /// @dev This does not include NXM from manager rewards (use stakingViewer.getManagerTokenRewards)
+  /// @param member The address of the member to query.
+  /// @param tokenIds An array of staking NFT token IDs associated with the member.
+  /// @return A ClaimableNxm struct containing details of the claimable NXM tokens.
+  function getClaimableNxm(address member, uint[] calldata tokenIds) public view returns (ClaimableNxm memory) {
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:50]: make sure to add the missing tokens computations, as pe comments above

File: contracts/modules/viewer/NexusViewer.sol [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+    IStakingViewer.AggregatedTokens memory aggregatedTokens = stakingViewer.getAggregatedTokens(tokenIds);
+
+    uint assessmentStakeAmount = 0;
+    if (assessmentViewer.isStakeLocked(member)) {
+      (uint96 amount, , ) = _assessment().stakeOf(member);
+      assessmentStakeAmount = uint(amount);
+    }
+
+    return LockedNxm({aggregatedTokens: aggregatedTokens, assessmentStakeAmount: assessmentStakeAmount});
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:86]: make sure to add missing fields, as per comments above

File: test/fork/nexus-viewer.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    expect(claimableNxm.aggregateStakingTokens.totalExpiredStake).to.be.instanceOf(BigNumber);
+    expect(claimableNxm.aggregateStakingTokens.totalRewards).to.be.instanceOf(BigNumber);
+    expect(claimableNxm.assessmentRewards.totalPendingAmountInNXM).to.be.instanceOf(BigNumber);
+    expect(claimableNxm.assessmentRewards.withdrawableAmountInNXM).to.be.instanceOf(BigNumber);
+    expect(claimableNxm.assessmentRewards.withdrawableUntilIndex).to.be.instanceOf(BigNumber);
+    expect(claimableNxm.legacyPooledStake.deposit).to.be.instanceOf(BigNumber);
+    expect(claimableNxm.legacyPooledStake.reward).to.be.instanceOf(BigNumber);
+    expect(claimableNxm.v1CoverNotesAmount).to.be.instanceOf(BigNumber);
+  });
+});
```
Comment (roxdanila) [inline test/fork/nexus-viewer.js:60]: similar to the previous PR, let's kill fork tests in favour of unit/integration tests and check actual values

**Review by roxdanila — CHANGES_REQUESTED** [review]
See comments

File: contracts/modules/viewer/NexusViewer.sol (hunk 6 above)
Comment (rackstar) [reply contracts/modules/viewer/NexusViewer.sol:46]: from what I understood with my discussion with @mixplore we want to separate it so that we don't call the manager specific calls for non pool managers.

@mixplore can you confirm how the FE prefers to call getClaimableNxm (with or without manager rewards)?

Comment (roxdanila) [reply contracts/modules/viewer/NexusViewer.sol:46]: it's inaccurate to display total claimable NXM without including the manager rewards in the dashboard. the separation makes sense for individual staking related calls, but not for aggregate data.

Comment (mixplore) [reply contracts/modules/viewer/NexusViewer.sol:46]: @rackstar It should include the manager rewards if they exist aka the member is a pool manager

**Comment by rackstar** [comment]
ready for review. but marking as DRAFT as it should not  be merged until https://github.com/NexusMutual/smart-contracts/pull/1170 and https://github.com/NexusMutual/smart-contracts/pull/1164 are merged first

File: contracts/interfaces/INexusViewer.sol [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+interface INexusViewer {
+
+  struct ClaimableNXM {
+    uint governanceRewards; // Governance rewards in NXM
+    uint assessmentRewards; // Claimable assessment reward in NXM
+    uint assessmentStake; // Claimable assessment stake in NXM
+    uint stakingPoolTotalRewards; // Total staking pool rewards in NXM
+    uint stakingPoolTotalExpiredStake; // Total staking pool expired stake in NXM
+    uint managerTotalRewards; // Pool manager total staking rewards in NXM
+    uint legacyPooledStakeRewards; // Legacy pooled staking deposits in NXM
```
Comment (roxdanila) [inline contracts/interfaces/INexusViewer.sol:17]: nit: `// Legacy pooled staking rewards in NXM`

File: contracts/modules/viewer/NexusViewer.sol [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+  IStakingViewer public immutable stakingViewer;
+  IAssessmentViewer public immutable assessmentViewer;
+
+  constructor(INXMMaster _master, IStakingViewer _stakingViewer, IAssessmentViewer _assessmentViewer) {
+    master = _master;
+    stakingViewer = _stakingViewer;
+    assessmentViewer = _assessmentViewer;
+  }
+
+  /// @notice Retrieves the claimable NXM tokens across the platform for a given member.
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:29]: platform -> protocol

File: contracts/modules/viewer/NexusViewer.sol [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+  function _getManagerTotalRewards(address member) internal view returns (uint managerTotalRewards) {
+
+    IStakingViewer.Token[] memory tokenRewards = stakingViewer.getManagerTokenRewards(member);
+
+    for (uint i = 0; i < tokenRewards.length; i++) {
+      managerTotalRewards += tokenRewards[i].rewards;
+    }
+  }
+
+  function _getAssessmentStake(address member) internal view returns (uint assessmentStake, bool isStakeLocked) {
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:95]: Should you move this to assessment viewer?

File: contracts/modules/viewer/NexusViewer.sol [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+    (uint assessmentStake, bool isStakeLocked) = _getAssessmentStake(member);
+
+    return StakedNXM({
+      stakingPoolTotalActiveStake: aggregatedTokens.totalActiveStake,
+      assessmentStake: isStakeLocked ? assessmentStake : 0,
+      assessmentRewards: assessmentRewards.totalPendingAmountInNXM - assessmentRewards.withdrawableAmountInNXM
+    });
+  }
+
+  function _getManagerTotalRewards(address member) internal view returns (uint managerTotalRewards) {
```
Comment (roxdanila) [inline contracts/modules/viewer/NexusViewer.sol:86]: Should you move this to StakingViewer? Might be useful more widely as a public function

File: contracts/interfaces/INexusViewer.sol (hunk 14 above)
Comment (rackstar) [reply contracts/interfaces/INexusViewer.sol:17]: oh good catch 👍

File: contracts/modules/viewer/NexusViewer.sol (hunk 17 above)
Comment (rackstar) [reply contracts/modules/viewer/NexusViewer.sol:86]: yes, its better that way. good suggestion 👍

File: contracts/interfaces/IStakingViewer.sol [hunk 18]
```diff
@@ -93,6 +94,8 @@ interface IStakingViewer {
 
   function getManagerTokenRewards(address manager) external view returns (Token[] memory tokens);
 
+  function getManagerTotalRewards(address member) external view returns (uint managerTotalRewards);
+
   function getManagerPoolsAndRewards(address manager) external view returns (ManagerPoolsAndRewards memory);
 
   function getManagerRewards(uint[] memory poolIds) external view returns (Token[] memory tokens);
```
Comment (roxdanila) [inline contracts/interfaces/IStakingViewer.sol:101]: 1. please stay consistent with naming the function argument: either `manager` or `member` across these functions
2. Slight scope creep: `getManagerTokenRewards` & `getManagerRewards` could be renamed to something a bit more descriptive: `getManagerTokenRewardsByAddr`, `getManagerTokenRewardsByPools`,

Comment (rackstar) [reply contracts/interfaces/IStakingViewer.sol:101]: i'll change to `getManagerTokenRewardsByAddr` but we'll need to keep `getManagerRewards` to avoid a breaking change as its an existing method

**Review by roxdanila — APPROVED** [review]
